c2rust-analyze/src/rewrite/mod.rs

```diff
@@ -0,0 +1,231 @@
+use rustc_hir::Mutability;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+use std::fmt;
+
+mod apply;
+mod expr;
+mod span_index;
+mod ty;
+
+pub use self::expr::gen_expr_rewrites;
+pub use self::ty::dump_rewritten_local_tys;
+pub use self::ty::gen_ty_rewrites;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Rewrite<S = Span> {
+    /// Take the original expression unchanged.
+    Identity,
+    /// Extract the subexpression at the given index.
+    Sub(usize, S),
+
+    // Expression builders
+    /// `&e`, `&mut e`
+    Ref(Box<Rewrite>, Mutability),
+    /// `core::ptr::addr_of!(e)`, `core::ptr::addr_of_mut!(e)`
+    AddrOf(Box<Rewrite>, Mutability),
+    /// `*e`
+    Deref(Box<Rewrite>),
+    /// `arr[idx]`
+    Index(Box<Rewrite>, Box<Rewrite>),
+    /// `arr[idx..]`
+    SliceTail(Box<Rewrite>, Box<Rewrite>),
+    /// `e as usize`
+    CastUsize(Box<Rewrite>),
+    /// The integer literal `0`.
+    LitZero,
+
+    // Type builders
+    /// Emit a complete pretty-printed type, discarding the original annotation.
+    PrintTy(String),
```

Comment:
why are expression rewrites and type rewrites part of the same enum?

Reply:
Mainly so we can apply all rewrites in a single pass.  Doing e.g. all the expressions in one pass and all the types in a second pass would be significantly more complicated.  To do it in a single pass, it's easier to use a single type for all rewrites because in the tree of rewrites built in `rewrite::apply`, an expr-rewrite node might have a ty-rewrite as a child, and (in theory) the reverse is possible as well.